lib/puppet/network/http/connection.rb

```diff
@@ -222,15 +222,15 @@ def with_connection(site, &block)
       # can be nil
       peer_cert = @verify.peer_certs.last
 
-      if peer_cert && !OpenSSL::SSL.verify_certificate_identity(peer_cert.content, site.host)
-        valid_certnames = [peer_cert.name, *peer_cert.subject_alt_names].uniq
-        msg = valid_certnames.length > 1 ? _("one of %{certnames}") % { certnames: valid_certnames.join(', ') } : valid_certnames.first
-        msg += _("Server hostname '%{host}' did not match server certificate; expected %{msg}") % { host: site.host, msg: msg }
-        raise Puppet::Error, msg, error.backtrace
-      elsif error.message.include? "certificate verify failed"
+      if error.message.include? "certificate verify failed"
         msg = error.message
         msg << ": [" + @verify.verify_errors.join('; ') + "]"
         raise Puppet::Error, msg, error.backtrace
+      elsif peer_cert && !OpenSSL::SSL.verify_certificate_identity(peer_cert.content, site.host)
+        valid_certnames = [peer_cert.name, *peer_cert.subject_alt_names].uniq
+        certnames_msg = valid_certnames.length > 1 ? _("one of %{certnames}") % { certnames: valid_certnames.join(', ') } : valid_certnames.first
+        msg = _("Server hostname '%{host}' did not match server certificate; expected %{msg}") % { host: site.host, msg: certnames_msg }
```

Comment:
This kind of message that we piece together this way is very hard to translate. I think the clearest way to clean this up would be to include the "expected" in the smaller strings above, someting like:
```valid_certnames.length > 1 ? _("expected one of %{certnames}") % { certnames: valid_certnames.join(', ') } : _("expected %{certname}") % { certname: valid_certnames.first }
```

Reply:
Done.